src/Common/FrequencyHolder.h

```diff
@@ -0,0 +1,237 @@
+#pragma once
+#include <Common/StringUtils/StringUtils.h>
+#include <IO/ReadBufferFromFile.h>
+#include <IO/ReadBufferFromString.h>
+#include <IO/ReadHelpers.h>
+#include <IO/readFloatText.h>
+#include <IO/Operators.h>
+#include <IO/ZstdInflatingReadBuffer.h>
+
+#include <Common/Arena.h>
+#include <base/StringRef.h>
+#include <Common/HashTable/HashMap.h>
+
+#include <string_view>
+#include <string>
+#include <cstring>
+#include <unordered_map>
+#include <base/logger_useful.h>
+#include <Common/getResource.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int FILE_DOESNT_EXIST;
+}
+
+class FrequencyHolder
+{
+
+public:
+    struct Language
+    {
+        String name;
+        HashMap<StringRef, Float64> map;
+    };
+
+    struct Encoding
+    {
+        String name;
+        HashMap<UInt16, Float64> map;
+    };
+
+public:
+    using Map = HashMap<StringRef, Float64>;
+    using Container = std::vector<Language>;
+    using EncodingMap = HashMap<UInt16, Float64>;
+    using EncodingContainer = std::vector<Encoding>;
+
+    static FrequencyHolder & getInstance()
+    {
+        static FrequencyHolder instance;
+        return instance;
+    }
+
+
+    void loadEncodingsFrequency()
+    {
+        Poco::Logger * log = &Poco::Logger::get("EncodingsFrequency");
+
+        LOG_TRACE(log, "Loading embedded charset frequencies");
+
+        auto resource = getResource("charset_freq.txt.zst");
+            if (resource.empty())
+                throw Exception(ErrorCodes::FILE_DOESNT_EXIST, "There is no embedded charset frequencies");
+
+        String line;
+        UInt16 bigram;
+        Float64 frequency;
+        String charset_name;
+
+        auto buf = std::make_unique<ReadBufferFromMemory>(resource.data(), resource.size());
+        std::unique_ptr<ReadBuffer> in = std::make_unique<ZstdInflatingReadBuffer>(std::move(buf));
+
+        while (!in->eof())
+        {
+            readString(line, *in);
+            ++in->position();
+
+            if (line.empty())
+                continue;
+
+            ReadBufferFromString buf_line(line);
```

Comment:
ReadBuffer on top of string read from another ReadBuffer scares me a little bit.
It is some external format?
If not, maybe we can avoid it?

Reply:
We need to read the full line in case the buffer ends somewhere in the center of the line (which is not good and may break everything). After that ReadBufferFromString is just a simple way to extract values out of it. If there is a better way to do it I would be happy to hear it :)